// @carlosnunezmx/animeflv/dist/index.js
"use strict";
/***
 *  @author CarlosNunezMX
 *  @license MIT
 *  @package @carlosnunezmx/animeflv
 * **/
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAnimeInfo = exports.GetNewEpisodes = exports.GetNewSeries = exports.GetHome = exports.GetResources = exports.Emmision = exports.Search = void 0;
var search_1 = require("./scrappers/search"); // Obtener resultados de browse
Object.defineProperty(exports, "Search", { enumerable: true, get: function () { return search_1.Search; } });
Object.defineProperty(exports, "Emmision", { enumerable: true, get: function () { return search_1.Emmision; } });
var episode_1 = require("./scrappers/episode"); // Obtener servidores de episodio
Object.defineProperty(exports, "GetResources", { enumerable: true, get: function () { return episode_1.GetResources; } });
var main_page_1 = require("./scrappers/main_page");
Object.defineProperty(exports, "GetHome", { enumerable: true, get: function () { return main_page_1.GetHome; } });
Object.defineProperty(exports, "GetNewSeries", { enumerable: true, get: function () { return main_page_1.GetNewSeries; } });
Object.defineProperty(exports, "GetNewEpisodes", { enumerable: true, get: function () { return main_page_1.GetNewEpisodes; } });
var anime_1 = require("./scrappers/anime");
Object.defineProperty(exports, "GetAnimeInfo", { enumerable: true, get: function () { return anime_1.GetAnimeInfo; } });
